### ISSUE TITLE
make it compatible with WMCore 1.2.8 (pycurl)

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#5984]
with recent changes to
https://github.com/dmwm/WMCore/blob/2ec701cd7785545663461ceb536e462deee0ecfa/src/python/WMCore/Services/pycurl_manager.py

we need to explicitely state that parameters for the URL need to be URLEncoded 

See:
https://github.com/dmwm/CRABServer/issues/5945#issuecomment-553167217
and
https://github.com/dmwm/WMCore/pull/9360#issuecomment-553163931

**Comment by belforte** [comment]
I have an initial fix, but it requires more testing to make sure that it works for all places/ways in which we use pycurl

**Comment by belforte** [comment]
See
https://github.com/dmwm/WMCore/pull/9360#issuecomment-553260004
and following comments

**Comment by belforte** [comment]
in particular Alan pointed out this
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/Requests.py#L153-L170
while CRAB the historical decision was not to use `WMCore/Services/Requests.py` but the simplified version https://github.com/belforte/CRABServer/blob/6090b1fa3dc279afe5c7f44f90ced9895f888edf/src/python/RESTInteractions.py#L35-L47

**Comment by amaltaro** [comment]
Stefano, this is how the pycurl module used to be a couple of months ago:
https://github.com/dmwm/WMCore/blob/e5c5f0e583221137b427e900b72285119c44c9bc/src/python/WMCore/Services/pycurl_manager.py#L115-L128

so if you have byte string parameters, you don't need any encoding.
Otherwise you need either urlencode or json.dumps, according to the HTTP method.

**Comment by belforte** [comment]
THanks Alan. And this is what I put in last night and worked in a first test
```python
        if verb is 'GET' and isinstance(data,dict):
            encode = True
        else:
            encode = False
        response, datares = self['conn'].request(url, data, encode=encode, headers=headers, verb=verb, doseq = True,
                                                 ckey=self['key'], cert=self['cert'], capath=caCertPath,
                                                 verbose=self['verbose'])
```
I wanted to know if this can be enough or if I need to make sure that I test all possible CRAB API's.
Just do not ask "why don't you guys have unit tests", please...

Anyhow I agree that I should start by inserting the very same code you had, so that I convert from dict to json for POST (at least). Onward...

**Comment by belforte** [comment]
Alan, same code as you pointed out is still there in the same place in the new `pycurl_manager`
https://github.com/dmwm/WMCore/blob/b538770e324426b9e8617d109c62db6b55bd8887/src/python/WMCore/Services/pycurl_manager.py#L117-L141
It looks to me that all that you did is to allow for the caller to say "don't touch it" which was previously only happening when params is a string.
So I would think that if I simply set Encode=True at all times, I get same behavior back.


Am I missing something ?
The pycurl_manager/RequestHandler/requests did not have the Encode argument
at the time of the code you pointed to, but it looks like it was behaving like it had.
Now you needed to change to accomodate PhEDEx et al. So be  it.
Simply I did not know!

https://github.com/dmwm/WMCore/blob/e5c5f0e583221137b427e900b72285119c44c9bc/src/python/WMCore/Services/pycurl_manager.py#L225-L227

And  in the end I did not know because you did not know that I needed to know !!!!
(let's take that offline... It is all Marco's fault. Tell him to buy you a couple of beers, since he can't buy one to me ! )

**Comment by belforte** [comment]
since I am at it, after fix is validated, I spotted a few things to cleanup in RESTInteractions.py
- [ ] buildUrl is not used anywere, can go away with urlunparse
- [ ] xrange should use range instead (py3 compat)
- [ ] at the end  of makeRequest datares and response are undefined if all tries fail, leading to odd exceptions
- [ ] decodeJson is only used here and should better be inlined in the call above

**Comment by belforte** [comment]
Good news:
 https://github.com/dmwm/CRABServer/blob/master/src/python/RESTInteractions.py is not used in CRABServer. No need to test this change inside cmsweb.
But it affects everything on the internal side: TaskWorker, Publisher, PostJob, DDMRequests, Rucio (TransferInterface) **AND** CRAB Client.

**Comment by belforte** [comment]
So need to build and test CRABClient as well with new WMCore

**Comment by belforte** [comment]
listing what I tested
- [ ] submit
- [ ] status
- [ ] resubmit
- [ ] kill
- [ ] get log
- [ ] get output
- [ ] report

**Comment by belforte** [comment]
solved via #5985

**Comment by amaltaro** [comment]
> It looks to me that all that you did is to allow for the caller to say "don't touch it" which was previously only happening when params is a string.
So I would think that if I simply set Encode=True at all times, I get same behavior back

That seems to be a fair statement! Besides all the additional handling at the upper classes, the pycurl had only these changes:
https://github.com/dmwm/WMCore/commit/4181fef43d5147f223fceca3e7a777524c0005cd#diff-1120a2d619920d0b79a217f9ee2b164e

which are resumed to:
* accept-encoding header support
* and accept a `encode` flag which can be used to not further encode anything at pycurl level

I'm gladly you got it fixed! Cheers!
